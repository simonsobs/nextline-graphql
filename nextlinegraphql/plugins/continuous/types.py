from asyncio import Task
from dataclasses import dataclass, field
from typing import Set

from nextline.utils.pubsub import PubSubItem


@dataclass
class ContinuousInfo:
    pubsub_enabled: PubSubItem[bool] = field(default_factory=PubSubItem)
    tasks: Set[Task] = field(default_factory=set)
