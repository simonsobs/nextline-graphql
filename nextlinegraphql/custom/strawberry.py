'''Strawberry GraphQL ASGI app with a fix for async_asgi_testclient.

Strawberry: https://strawberry.rocks
'''
from strawberry.asgi import GraphQL as GraphQL_
from strawberry.subscriptions import GRAPHQL_WS_PROTOCOL


class GraphQL(GraphQL_):
    """Add a fix to the strawberry GraphQL for async_asgi_testclient"""

    async def __call__(self, scope, receive, send):
        if scope["type"] == "websocket":
            if not scope.get("subprotocols"):
                # strawberry closes the websocket connection if
                # subprotocols are empty, which is the case for
                # async_asgi_testclient.
                scope["subprotocols"] = [GRAPHQL_WS_PROTOCOL]
        await super().__call__(scope, receive, send)
