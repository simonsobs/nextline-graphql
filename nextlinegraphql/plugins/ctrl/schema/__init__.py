__all__ = ['Query', 'Mutation', 'Subscription']

from .mutation import Mutation
from .query import Query
from .subscription import Subscription
