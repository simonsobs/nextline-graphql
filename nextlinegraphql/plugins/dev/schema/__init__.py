__all__ = ['Query']

from .query import Query

Mutation = None
Subscription = None
