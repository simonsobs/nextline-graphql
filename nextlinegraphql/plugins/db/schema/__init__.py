__all__ = ['Query', 'Mutation', 'Subscription']
import strawberry

from .query import Query


@strawberry.type
class Mutation:
    pass


@strawberry.type
class Subscription:
    pass
