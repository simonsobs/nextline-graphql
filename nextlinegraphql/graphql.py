from typing import Any, Optional

import strawberry
from nextline import Nextline
from strawberry.schema import BaseSchema
from strawberry.tools import merge_types

from .custom.pluggy import PluginManager
from .custom.strawberry import GraphQL


def compose_schema(hook: PluginManager) -> BaseSchema:

    # [(Query, Mutation, Subscription), ...]
    three_types = hook.hook.schema()

    # [(Query, ...), (Mutation, ...), (Subscription, ...)]
    transposed = list(map(tuple, zip(*three_types)))
    transposed = [tuple(t for t in l if t) for l in transposed]  # remove None

    q, m, s = transposed
    assert q  # Query is required
    Query = merge_types('Query', q)  # type: ignore
    Mutation = m and merge_types('Mutation', m) or None  # type: ignore
    Subscription = s and merge_types('Subscription', s) or None  # type: ignore

    schema = strawberry.Schema(
        query=Query,
        mutation=Mutation,
        subscription=Subscription,
    )

    return schema


class EGraphQL(GraphQL):
    """Extend the strawberry GraphQL app

    https://strawberry.rocks/docs/integrations/asgi
    """

    def __init__(
        self,
        schema: BaseSchema,
        nextline: Nextline,
        hook: PluginManager,
    ):
        super().__init__(schema)
        self._nextline = nextline
        self._hook = hook

    async def get_context(self, request, response=None) -> Optional[Any]:
        context = {
            "request": request,
            "response": response,
            "nextline": self._nextline,
        }
        updates = await self._hook.ahook.get_context(context=context)
        for update in updates:
            if update:
                context.update(update)
        return context
