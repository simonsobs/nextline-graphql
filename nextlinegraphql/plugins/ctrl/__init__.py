__all__ = ['Plugin']

from collections.abc import AsyncIterator, MutableMapping

import apluggy as pluggy
from apluggy import asynccontextmanager
from nextline import Nextline

from nextlinegraphql.hook import spec

from .example_script import statement
from .schema import Mutation, Query, Subscription


class Plugin:
    @spec.hookimpl
    def schema(self):
        return (Query, Mutation, Subscription)

    @spec.hookimpl
    async def update_lifespan_context(
        self, hook: pluggy.PluginManager, context: MutableMapping
    ) -> None:
        run_no: int = max(hook.hook.initial_run_no(), default=1)
        script: str = [*hook.hook.initial_script(), statement][0]
        self._nextline = Nextline(script, run_no)
        context['nextline'] = self._nextline

    @spec.hookimpl(trylast=True)  # trylast so to be the innermost context
    @asynccontextmanager
    async def lifespan(self) -> AsyncIterator[None]:
        '''Yield within the nextline context.'''
        async with self._nextline:
            yield

    @spec.hookimpl
    def update_strawberry_context(self, context: MutableMapping) -> None:
        context['nextline'] = self._nextline
