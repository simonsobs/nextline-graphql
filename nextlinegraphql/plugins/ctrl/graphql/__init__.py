from pathlib import Path

from graphql import parse, print_ast


def read_gql(path: Path | str) -> str:
    '''Load a GraphQL query from a file while checking its syntax.'''

    text = Path(path).read_text()
    parsed = parse(text)
    reformatted = print_ast(parsed)
    return reformatted


pwd = Path(__file__).resolve().parent

sub = pwd / 'mutations'
MUTATE_EXEC = read_gql(sub / 'Exec.gql')
MUTATE_RUN_AND_CONTINUE = read_gql(sub / 'RunAndContinue.gql')
MUTATE_RESET = read_gql(sub / 'Reset.gql')
MUTATE_SEND_PDB_COMMAND = read_gql(sub / 'SendPdbCommand.gql')
MUTATE_INTERRUPT = read_gql(sub / 'Interrupt.gql')
MUTATE_TERMINATE = read_gql(sub / 'Terminate.gql')
MUTATE_KILL = read_gql(sub / 'Kill.gql')
MUTATE_LOAD_EXAMPLE_SCRIPT = read_gql(sub / 'LoadExampleScript.gql')

sub = pwd / 'queries'
QUERY_STATE = read_gql(sub / 'State.gql')
QUERY_RUN_NO = read_gql(sub / 'RunNo.gql')
QUERY_TRACE_IDS = read_gql(sub / 'TraceIds.gql')
QUERY_SOURCE = read_gql(sub / 'Source.gql')
QUERY_SOURCE_LINE = read_gql(sub / 'SourceLine.gql')
QUERY_EXCEPTION = read_gql(sub / 'Exception.gql')
QUERY_CONTINUOUS_ENABLED = read_gql(sub / 'ContinuousEnabled.gql')

sub = pwd / 'subscriptions'
SUBSCRIBE_COUNTER = read_gql(sub / 'Counter.gql')
SUBSCRIBE_STATE = read_gql(sub / 'State.gql')
SUBSCRIBE_RUN_NO = read_gql(sub / 'RunNo.gql')
SUBSCRIBE_TRACE_IDS = read_gql(sub / 'TraceIds.gql')
SUBSCRIBE_PROMPTING = read_gql(sub / 'Prompting.gql')
SUBSCRIBE_STDOUT = read_gql(sub / 'Stdout.gql')
SUBSCRIBE_CONTINUOUS_ENABLED = read_gql(sub / 'ContinuousEnabled.gql')
