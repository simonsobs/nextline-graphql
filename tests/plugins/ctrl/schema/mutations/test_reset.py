from pathlib import Path

import pytest
from nextline import Nextline
from strawberry.types import ExecutionResult

from nextlinegraphql.plugins.ctrl import example_script as example_script_module
from nextlinegraphql.plugins.ctrl.graphql import MUTATE_RESET, QUERY_SOURCE
from tests.plugins.ctrl.schema.conftest import Schema

EXAMPLE_SCRIPT_PATH = Path(example_script_module.__file__).parent / 'script.py'
example_script = EXAMPLE_SCRIPT_PATH.read_text()

SOURCE_ONE = '''
import time
time.sleep(0.1)
'''.strip()

params = [
    pytest.param(None, id='no-statement'),
    pytest.param(SOURCE_ONE, id='statement'),
]


@pytest.mark.parametrize('statement', params)
async def test_schema(schema: Schema, statement: str | None) -> None:
    nextline = Nextline(example_script, trace_modules=True, trace_threads=True)
    async with nextline:
        context = {'nextline': nextline}
        variables = {'statement': statement}
        result = await schema.execute(
            MUTATE_RESET, context_value=context, variable_values=variables
        )
        assert isinstance(result, ExecutionResult)
        assert (data := result.data)
        assert data['ctrl']['reset'] is True

        result = await schema.execute(QUERY_SOURCE, context_value=context)
        assert isinstance(result, ExecutionResult)
        assert (data := result.data)

        expected = statement or example_script
        assert expected.split('\n') == data['ctrl']['source']
