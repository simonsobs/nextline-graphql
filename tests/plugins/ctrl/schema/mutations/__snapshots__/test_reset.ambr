# serializer version: 1
# name: test_reset[no-statement]
  dict({
    'data': dict({
      'source': list([
        'import time',
        '',
        'time.sleep(0.3)',
        '',
        'def f():',
        '    for _ in range(10):',
        '        pass',
        '    return',
        '',
        'f()',
        'f()',
        '',
        "print('here!')",
        '',
        'import script_threading',
        '',
        'script_threading.run()',
        '',
        '',
        'import script_asyncio',
        '',
        'script_asyncio.run()',
        '',
      ]),
    }),
  })
# ---
# name: test_reset[statement]
  dict({
    'data': dict({
      'source': list([
        'import time',
        'time.sleep(0.1)',
      ]),
    }),
  })
# ---
